### PR TITLE
Add CliArgs to config

### PR DESCRIPTION
**Pull request by tinchou**
This PR basically adds `CliArgs` as a source to the config. We can now pass the following:

```
--regtest, --testnet:   network to connect to, default is mainnet
--reset:                reset the current database
-rpccors [value]:       enables RPC and uses [value] for CORS validation
```

It also simplifies some configuration:

* Now `ConfigLoader` has full responsibility over reading and overlaying multiple configuration sources, and it was refactored to be clearer
* Removes the old `CLIInterface` and uses our own `CliArgs` set of classes, which provide strongly-typed argument parsing
* Creates a `NodeBootstrapper` class so we can start decoupling the initial object graph creation

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 76 issues
* ![MAJOR][MAJOR] 47 major
* ![MINOR][MINOR] 12 minor
* ![INFO][INFO] 17 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [DoPrune.java#L72](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/DoPrune.java#L72): Remove this unused method parameter "args". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [DoPrune.java#L72](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/DoPrune.java#L72): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [Start.java#L66](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/Start.java#L66): Remove this unused "txHandler" private field. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [Start.java#L81](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/Start.java#L81): Constructor has 16 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [Start.java#L115](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/Start.java#L115): The Cyclomatic Complexity of this method "startNode" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Start.java#L115](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/Start.java#L115): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [CliArgs.java#L81](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L81): Rename "options" which hides the field declared at line 71. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [CliArgs.java#L82](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L82): Rename "flags" which hides the field declared at line 72. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [CliArgs.java#L86](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L86): Reduce this switch case number of lines from 18 to at most 5, for example by extracting code into methods. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1151)
1. ![MAJOR][MAJOR] [CliArgs.java#L91](https://github.com/rsksmart/rskj/blob/fdedefd81ac25e7e0d84551831ef27386b20fbf1/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L91): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/NodeBootstrapper.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import co.rsk.config.NodeCliFlags;
+import co.rsk.config.NodeCliOptions;
+import org.ethereum.config.DefaultConfig;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+/**
+ * Bootstraps the node, creating the initial object graph.
+ * It uses Spring at the moment.
+ */
+class NodeBootstrapper {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/NodeBootstrapper.java]: I think it's better to use `public`/`private` explicitly instead of the _default_ visibility

**Comment by rskops** [comment]
SonarQube analysis reported 76 issues
* ![MAJOR][MAJOR] 47 major
* ![MINOR][MINOR] 11 minor
* ![INFO][INFO] 18 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [DoPrune.java#L67](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/DoPrune.java#L67): Class co.rsk.DoPrune defines fields that are used only as locals [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AFCBL_FIELD_COULD_BE_LOCAL)
1. ![MAJOR][MAJOR] [DoPrune.java#L72](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/DoPrune.java#L72): Remove this unused method parameter "args". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [DoPrune.java#L72](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/DoPrune.java#L72): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [FullNodeRunner.java#L73](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/FullNodeRunner.java#L73): Constructor has 16 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [FullNodeRunner.java#L109](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/FullNodeRunner.java#L109): The Cyclomatic Complexity of this method "run" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [NodeRunner.java#L24](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/NodeRunner.java#L24): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [CliArgs.java#L81](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L81): Rename "options" which hides the field declared at line 71. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [CliArgs.java#L82](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L82): Rename "flags" which hides the field declared at line 72. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [CliArgs.java#L86](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L86): Reduce this switch case number of lines from 18 to at most 5, for example by extracting code into methods. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1151)
1. ![MAJOR][MAJOR] [CliArgs.java#L91](https://github.com/rsksmart/rskj/blob/aa34906f3e8b93b22e5054717679a8bfc8895cec/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L91): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by tinchou** [comment]
The following command line parameters are available:

```
--regtest, --testnet:   network to connect to, default is mainnet
--reset:                reset the current database
-rpccors [value]:       enables RPC and sets [value] for CORS
```

Does this go in the wiki?

**Comment by rskops** [comment]
SonarQube analysis reported 78 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 24 major
* ![MINOR][MINOR] 33 minor
* ![INFO][INFO] 18 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [DoPrune.java#L73](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/DoPrune.java#L73): Remove this unused method parameter "args". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1172)
1. ![CRITICAL][CRITICAL] [SystemProperties.java#L462](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L462): Method org.ethereum.config.SystemProperties.getGeneratedNodePrivateKey() passes a concatenated string to SLF4J's format string [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![CRITICAL][CRITICAL] [SystemProperties.java#L463](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java#L463): Method org.ethereum.config.SystemProperties.getGeneratedNodePrivateKey() passes a concatenated string to SLF4J's format string [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![MAJOR][MAJOR] [FullNodeRunner.java#L109](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/FullNodeRunner.java#L109): The Cyclomatic Complexity of this method "run" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [CliArgs.java#L86](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L86): Reduce this switch case number of lines from 18 to at most 5, for example by extracting code into methods. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1151)
1. ![MAJOR][MAJOR] [CliArgs.java#L91](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L91): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [CliArgs.java#L96](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L96): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [CliArgs.java#L102](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L102): Refactor the code in order to not assign to this loop counter from within the loop body. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)
1. ![MAJOR][MAJOR] [MigrationTool.java#L40](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L40): Method co.rsk.cli.config.MigrationTool.main(String[]) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [MigrationTool.java#L50](https://github.com/rsksmart/rskj/blob/b62d2ed6ed79013ab748564946c2af939508fa8d/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L50): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 73 issues
* ![MAJOR][MAJOR] 24 major
* ![MINOR][MINOR] 31 minor
* ![INFO][INFO] 18 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [FullNodeRunner.java#L109](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/FullNodeRunner.java#L109): The Cyclomatic Complexity of this method "run" is 12 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [CliArgs.java#L86](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L86): Reduce this switch case number of lines from 18 to at most 5, for example by extracting code into methods. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1151)
1. ![MAJOR][MAJOR] [CliArgs.java#L91](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L91): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [CliArgs.java#L96](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L96): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [CliArgs.java#L102](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/CliArgs.java#L102): Refactor the code in order to not assign to this loop counter from within the loop body. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)
1. ![MAJOR][MAJOR] [MigrationTool.java#L40](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L40): Method co.rsk.cli.config.MigrationTool.main(String[]) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [MigrationTool.java#L50](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L50): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [MigrationTool.java#L51](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L51): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [MigrationTool.java#L51](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/cli/config/MigrationTool.java#L51): Format string should use %n rather than \n in co.rsk.cli.config.MigrationTool.main(String[]) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3AVA_FORMAT_STRING_USES_NEWLINE)
1. ![MAJOR][MAJOR] [NodeCliFlags.java#L30](https://github.com/rsksmart/rskj/blob/869597fb792a64c3ec28cb3b0e893c4b701d27a7/rskj-core/src/main/java/co/rsk/config/NodeCliFlags.java#L30): Method co.rsk.config.NodeCliFlags.<static initializer for NodeCliFlags>() needlessly boxes a boolean constant [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ANAB_NEEDLESS_BOOLEAN_CONSTANT_CONVERSION)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**